crates/sui-json-rpc/src/error.rs

```diff
@@ -84,34 +85,133 @@ impl Error {
                 }
                 _ => RpcError::Call(CallError::Failed(err.into())),
             },
+            Error::ClientError(err) => err.into(),
+            Error::ServerError(err) => err.into(),
             _ => RpcError::Call(CallError::Failed(self.into())),
         }
     }
 }
 
 #[derive(Debug, Error)]
-pub enum SuiRpcInputError {
-    #[error("Input contains duplicates")]
-    ContainsDuplicates,
+pub enum ServerError {
+    #[error("Serde error")]
+    Serde(SuiError),
 
-    #[error("Input exceeds limit of {0}")]
-    SizeLimitExceeded(String),
+    #[error("Bcs error")]
+    Bcs(bcs::Error),
+}
 
-    #[error("{0}")]
-    GenericNotFound(String),
+impl From<ServerError> for RpcError {
+    fn from(_e: ServerError) -> Self {
+        to_internal_error("Internal server error, please try again later")
+    }
+}
 
-    #[error("{0}")]
-    GenericInvalid(String),
+fn to_internal_error(err: impl ToString) -> RpcError {
+    let error_object = ErrorObject::owned(-32603, err.to_string(), None::<String>);
+    RpcError::Call(CallError::Custom(error_object))
+}
+
+// One-to-one of the return type of the RPC method where applicable
+#[derive(Debug, Clone, Copy, AsRefStr, EnumString)]
+pub enum EntityType {
+    DelegatedStake,
+    SuiMoveNormalizedStruct,
+    SuiMoveNormalizedFunction,
+    Object,
+    Package,
+    Product,
+    SuiMoveNormalizedModule,
+    Checkpoint,
+}
+
+impl fmt::Display for EntityType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    // Any kind of serialization or deserialization error
+    #[error("Serialization error on '{param}': {reason}")]
+    Serde { param: &'static str, reason: String },
+
+    #[error("Base64 error on '{param}': {reason}")]
+    Base64 { param: &'static str, reason: String },
+
+    #[error("BCS error on '{param}': {reason}")]
+    Bcs {
+        param: &'static str,
+        reason: bcs::Error,
+    },
+
+    #[error("FastCrypto error on '{param}': {reason}")]
+    FastCrypto {
+        param: &'static str,
+        reason: FastCryptoError,
+    },
 
-    #[error("request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
+    #[error("Invalid combination of type and value: {0}")]
+    SerdeWithLayout(String),
+
+    #[error(transparent)]
+    Domain(#[from] DomainParseError),
+
+    #[error("Invalid '{param}': {reason}")]
+    InvalidParam { param: &'static str, reason: String },
+    // maybe InvalidParamMulti or something. param, value, reason
+    #[error("`request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
```

Comment:
```suggestion
    #[error("You must set `request_type` to `None` or `WaitForLocalExecution` if the response requires effects.")]
```

Seems like effects is a code element, so `effects`?